### PR TITLE
HHH-11171 Fixes for Java 8 time descriptors

### PR DESCRIPTION
**Pull request by coladict**
If you're worried that I'm changing InstantJavaDescriptor to use a different format, the original code never worked in the first place.

Tests included.

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/type/InstantType.java [hunk 1]
```diff
@@ -32,15 +33,16 @@
 	 */
 	public static final InstantType INSTANCE = new InstantType();
 
-	public static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern( "yyyy-MM-dd HH:mm:ss.S 'Z'", Locale.ENGLISH );
+	public static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern( "yyyy-MM-dd HH:mm:ss.n ZZZZ", Locale.ENGLISH );
```
Comment (vladmihalcea) [inline hibernate-core/src/main/java/org/hibernate/type/InstantType.java]: Why did you change from `S` (fractions of a second) to `n` (nanoseconds)?

Why `ZZZZ` (the long TimeZone format) instead of `Z`?

File: hibernate-core/src/main/java/org/hibernate/type/InstantType.java [hunk 2]
```diff
@@ -32,15 +33,16 @@
...
 
 	public InstantType() {
 		super( TimestampTypeDescriptor.INSTANCE, InstantJavaDescriptor.INSTANCE );
 	}
 
 	@Override
 	public String objectToSQLString(Instant value, Dialect dialect) throws Exception {
-		return "{ts '" + FORMATTER.format( ZonedDateTime.ofInstant( value, ZoneId.of( "UTC" ) ) ) + "'}";
+		// SQL 9075-2:2003 5.3 <timestamp literal>
+		return "TIMESTAMP '" + FORMATTER.format( ZonedDateTime.ofInstant( value, ZoneId.of( "UTC" ) ) ) + "'";
```
Comment (vladmihalcea) [inline hibernate-core/src/main/java/org/hibernate/type/InstantType.java]: That's a good catch. `{ts value}` is a ODBC escape syntax.

File: hibernate-core/src/main/java/org/hibernate/type/descriptor/java/InstantJavaDescriptor.java [hunk 3]
```diff
@@ -36,12 +36,12 @@ public InstantJavaDescriptor() {
 
 	@Override
 	public String toString(Instant value) {
-		return InstantType.FORMATTER.format( ZonedDateTime.ofInstant( value, ZoneId.of( "UTC" ) ) );
+		return DateTimeFormatter.ISO_INSTANT.format( value );
```
Comment (vladmihalcea) [inline hibernate-core/src/main/java/org/hibernate/type/descriptor/java/InstantJavaDescriptor.java]: `ISO_INSTANT` is also using UTC, so it's a good alternative to our previous formatting option.

**Comment by vladmihalcea** [comment]
Looks very good. I want to add a test for the `InstantType.objectToSQLString` as well so that we can verify the `FORMATTER` is working too.

File: hibernate-core/src/main/java/org/hibernate/type/InstantType.java (hunk 1 above)
Comment (coladict) [inline hibernate-core/src/main/java/org/hibernate/type/InstantType.java]: objectToSQLString is only called for values of Discriminator columns. I spent a few hours tracking down it's usage yesterday.

**Comment by vladmihalcea** [comment]
Merged it upstream. Thanks.
